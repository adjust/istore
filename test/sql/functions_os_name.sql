SELECT exist('ios=>1'::os_name_istore, 'ios');
SELECT exist('ios=>1'::os_name_istore, 'android');
SELECT exist('ios=>1, android=>0'::os_name_istore, 'ios');
SELECT exist('ios=>1, android=>0'::os_name_istore, 'windows');

SELECT fetchval('ios=>1'::os_name_istore, 'ios');
SELECT fetchval('windows=>1'::os_name_istore, 'ios');
SELECT fetchval('ios=>1, ios=>1'::os_name_istore, 'ios');
SELECT fetchval('ios=>1, ios=>1'::os_name_istore, 'windows');

SELECT add('ios=>1, windows=>1'::os_name_istore, 'ios=>1, windows=>1'::os_name_istore);
SELECT add('ios=>1, windows=>1'::os_name_istore, 'android=>1, windows=>1'::os_name_istore);
SELECT add('ios=>1, windows=>1'::os_name_istore, 'android=>-1, windows=>1'::os_name_istore);
SELECT add('android=>1, windows=>1'::os_name_istore, 'android=>-1, windows=>1'::os_name_istore);
SELECT add('android=>-1, windows=>1'::os_name_istore, 'android=>-1, windows=>1'::os_name_istore);
SELECT add('android=>-1, windows=>1'::os_name_istore, 1);
SELECT add('android=>-1, windows=>1'::os_name_istore, -1);
SELECT add('android=>-1, windows=>1'::os_name_istore, 0);

SELECT subtract('ios=>1, windows=>1'::os_name_istore, 'ios=>1, windows=>1'::os_name_istore);
SELECT subtract('ios=>1, windows=>1'::os_name_istore, 'android=>1, windows=>1'::os_name_istore);
SELECT subtract('ios=>1, windows=>1'::os_name_istore, 'android=>-1, windows=>1'::os_name_istore);
SELECT subtract('android=>1, windows=>1'::os_name_istore, 'android=>-1, windows=>1'::os_name_istore);
SELECT subtract('android=>-1, windows=>1'::os_name_istore, 'android=>-1, windows=>1'::os_name_istore);
SELECT subtract('android=>-1, windows=>1'::os_name_istore, 1);
SELECT subtract('android=>-1, windows=>1'::os_name_istore, -1);
SELECT subtract('android=>-1, windows=>1'::os_name_istore, 0);

SELECT multiply('ios=>1, windows=>1'::os_name_istore, 'ios=>1, windows=>1'::os_name_istore);
SELECT multiply('ios=>1, windows=>1'::os_name_istore, 'android=>1, windows=>1'::os_name_istore);
SELECT multiply('ios=>1, windows=>1'::os_name_istore, 'android=>-1, windows=>1'::os_name_istore);
SELECT multiply('android=>1, windows=>1'::os_name_istore, 'android=>-1, windows=>1'::os_name_istore);
SELECT multiply('android=>-1, windows=>1'::os_name_istore, 'android=>-1, windows=>1'::os_name_istore);
SELECT multiply('android=>-1, windows=>1'::os_name_istore, 1);
SELECT multiply('android=>-1, windows=>1'::os_name_istore, -1);
SELECT multiply('android=>-1, windows=>1'::os_name_istore, 0);

SELECT os_name_istore_from_array(ARRAY['android']);
SELECT os_name_istore_from_array(ARRAY['android','android','android','android']);
SELECT os_name_istore_from_array(NULL::text[]);
SELECT os_name_istore_from_array(ARRAY['android','ios','windows','windows-phone']);
SELECT os_name_istore_from_array(ARRAY['android','ios','windows','windows-phone','android','ios','windows','windows-phone']);
SELECT os_name_istore_from_array(ARRAY['android','ios','windows','windows-phone','android','ios','windows',NULL]);
SELECT os_name_istore_from_array(ARRAY[NULL,'ios','windows','windows-phone','android','ios','windows','windows-phone']);
SELECT os_name_istore_from_array(ARRAY[NULL,'ios','windows','windows-phone','android','ios','windows',NULL]);
SELECT os_name_istore_from_array(ARRAY['android','ios','windows',NULL,'android',NULL,'windows','windows-phone','android','ios','windows']);

SELECT os_name_istore_from_array(ARRAY['android'::os_name]);
SELECT os_name_istore_from_array(ARRAY['android'::os_name,'android'::os_name,'android'::os_name,'android'::os_name]);
SELECT os_name_istore_from_array(NULL::text[]);
SELECT os_name_istore_from_array(ARRAY['android'::os_name,'ios'::os_name,'windows'::os_name,'windows-phone'::os_name]);
SELECT os_name_istore_from_array(ARRAY['android'::os_name,'ios'::os_name,'windows'::os_name,'windows-phone'::os_name,'android'::os_name,'ios'::os_name,'windows'::os_name,'windows-phone'::os_name]);
SELECT os_name_istore_from_array(ARRAY['android'::os_name,'ios'::os_name,'windows'::os_name,'windows-phone'::os_name,'android'::os_name,'ios'::os_name,'windows'::os_name,NULL]);
SELECT os_name_istore_from_array(ARRAY[NULL,'ios'::os_name,'windows'::os_name,'windows-phone'::os_name,'android'::os_name,'ios'::os_name,'windows'::os_name,'windows-phone'::os_name]);
SELECT os_name_istore_from_array(ARRAY[NULL,'ios'::os_name,'windows'::os_name,'windows-phone'::os_name,'android'::os_name,'ios'::os_name,'windows'::os_name,NULL]);
SELECT os_name_istore_from_array(ARRAY['android'::os_name,'ios'::os_name,'windows'::os_name,NULL,'android'::os_name,NULL,'windows'::os_name,'windows-phone'::os_name,'android'::os_name,'ios'::os_name,'windows'::os_name]);

SELECT os_name_istore_agg(ARRAY['ios=>1']::os_name_istore[]);
SELECT os_name_istore_agg(ARRAY['ios=>1','ios=>1']::os_name_istore[]);
SELECT os_name_istore_agg(ARRAY['ios=>1,windows=>1','ios=>1,windows=>-1']::os_name_istore[]);
SELECT os_name_istore_agg(ARRAY['ios=>1,windows=>1','ios=>1,windows=>-1',NULL]::os_name_istore[]);
SELECT os_name_istore_agg(ARRAY[NULL,'ios=>1,windows=>1','ios=>1,windows=>-1']::os_name_istore[]);
SELECT os_name_istore_agg(ARRAY[NULL,'ios=>1,windows=>1','ios=>1,windows=>-1',NULL]::os_name_istore[]);

SELECT os_name_istore_sum_up('ios=>1'::os_name_istore);
SELECT os_name_istore_sum_up(NULL::os_name_istore);
SELECT os_name_istore_sum_up('ios=>1, windows=>1'::os_name_istore);
SELECT os_name_istore_sum_up('ios=>1 ,windows=>-1, ios=>1'::os_name_istore);

BEGIN;
CREATE TABLE test (a os_name_istore);
INSERT INTO test VALUES('ios=>1');
INSERT INTO test VALUES('windows=>1');
INSERT INTO test VALUES('windows-phone=>1');
SELECT SUM(a) FROM test;
ROLLBACK;

BEGIN;
CREATE TABLE test (a os_name_istore);
INSERT INTO test VALUES('ios=>1');
INSERT INTO test VALUES('windows=>1');
INSERT INTO test VALUES('windows-phone=>1');
INSERT INTO test VALUES(NULL);
INSERT INTO test VALUES('windows-phone=>3');
SELECT SUM(a) FROM test;
ROLLBACK;

