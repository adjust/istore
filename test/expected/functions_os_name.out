BEGIN;
-- functions_os_name should find presence of a key;
-- ./spec/functions_os_name_spec.rb:8;
CREATE EXTENSION istore;
SELECT exist('ios=>1'::os_name_istore, 'ios');
 exist 
-------
 t
(1 row)

SELECT exist('ios=>1'::os_name_istore, 'android');
 exist 
-------
 f
(1 row)

SELECT exist('ios=>1, android=>0'::os_name_istore, 'ios');
 exist 
-------
 t
(1 row)

SELECT exist('ios=>1, android=>0'::os_name_istore, 'windows');
 exist 
-------
 f
(1 row)

ROLLBACK;
BEGIN;
-- functions_os_name should fetch values;
-- ./spec/functions_os_name_spec.rb:15;
CREATE EXTENSION istore;
SELECT fetchval('ios=>1'::os_name_istore, 'ios');
 fetchval 
----------
        1
(1 row)

SELECT fetchval('windows=>1'::os_name_istore, 'ios');
 fetchval 
----------
         
(1 row)

SELECT fetchval('ios=>1, ios=>1'::os_name_istore, 'ios');
 fetchval 
----------
        2
(1 row)

SELECT fetchval('ios=>1, ios=>1'::os_name_istore, 'windows');
 fetchval 
----------
         
(1 row)

ROLLBACK;
BEGIN;
-- functions_os_name should add to os_name_istores;
-- ./spec/functions_os_name_spec.rb:22;
CREATE EXTENSION istore;
SELECT add('ios=>1, windows=>1'::os_name_istore, 'ios=>1, windows=>1'::os_name_istore);
            add            
---------------------------
 "ios"=>"2","windows"=>"2"
(1 row)

SELECT add('ios=>1, windows=>1'::os_name_istore, 'android=>1, windows=>1'::os_name_istore);
                   add                    
------------------------------------------
 "android"=>"1","ios"=>"1","windows"=>"2"
(1 row)

SELECT add('ios=>1, windows=>1'::os_name_istore, 'android=>-1, windows=>1'::os_name_istore);
                    add                    
-------------------------------------------
 "android"=>"-1","ios"=>"1","windows"=>"2"
(1 row)

SELECT add('android=>1, windows=>1'::os_name_istore, 'android=>-1, windows=>1'::os_name_istore);
              add              
-------------------------------
 "android"=>"0","windows"=>"2"
(1 row)

SELECT add('android=>-1, windows=>1'::os_name_istore, 'android=>-1, windows=>1'::os_name_istore);
              add               
--------------------------------
 "android"=>"-2","windows"=>"2"
(1 row)

SELECT add('android=>-1, windows=>1'::os_name_istore, 1);
              add              
-------------------------------
 "android"=>"0","windows"=>"2"
(1 row)

SELECT add('android=>-1, windows=>1'::os_name_istore, -1);
              add               
--------------------------------
 "android"=>"-2","windows"=>"0"
(1 row)

SELECT add('android=>-1, windows=>1'::os_name_istore, 0);
              add               
--------------------------------
 "android"=>"-1","windows"=>"1"
(1 row)

ROLLBACK;
BEGIN;
-- functions_os_name should substract to os_name_istores;
-- ./spec/functions_os_name_spec.rb:48;
CREATE EXTENSION istore;
SELECT subtract('ios=>1, windows=>1'::os_name_istore, 'ios=>1, windows=>1'::os_name_istore);
         subtract          
---------------------------
 "ios"=>"0","windows"=>"0"
(1 row)

SELECT subtract('ios=>1, windows=>1'::os_name_istore, 'android=>1, windows=>1'::os_name_istore);
                 subtract                  
-------------------------------------------
 "android"=>"-1","ios"=>"1","windows"=>"0"
(1 row)

SELECT subtract('ios=>1, windows=>1'::os_name_istore, 'android=>-1, windows=>1'::os_name_istore);
                 subtract                 
------------------------------------------
 "android"=>"1","ios"=>"1","windows"=>"0"
(1 row)

SELECT subtract('android=>1, windows=>1'::os_name_istore, 'android=>-1, windows=>1'::os_name_istore);
           subtract            
-------------------------------
 "android"=>"2","windows"=>"0"
(1 row)

SELECT subtract('android=>-1, windows=>1'::os_name_istore, 'android=>-1, windows=>1'::os_name_istore);
           subtract            
-------------------------------
 "android"=>"0","windows"=>"0"
(1 row)

SELECT subtract('android=>-1, windows=>1'::os_name_istore, 1);
            subtract            
--------------------------------
 "android"=>"-2","windows"=>"0"
(1 row)

SELECT subtract('android=>-1, windows=>1'::os_name_istore, -1);
           subtract            
-------------------------------
 "android"=>"0","windows"=>"2"
(1 row)

SELECT subtract('android=>-1, windows=>1'::os_name_istore, 0);
            subtract            
--------------------------------
 "android"=>"-1","windows"=>"1"
(1 row)

ROLLBACK;
BEGIN;
-- functions_os_name should multiply two os_name_istores;
-- ./spec/functions_os_name_spec.rb:74;
CREATE EXTENSION istore;
SELECT multiply('ios=>1, windows=>1'::os_name_istore, 'ios=>1, windows=>1'::os_name_istore);
         multiply          
---------------------------
 "ios"=>"1","windows"=>"1"
(1 row)

SELECT multiply('ios=>1, windows=>1'::os_name_istore, 'android=>1, windows=>1'::os_name_istore);
                  multiply                  
--------------------------------------------
 "android"=>NULL,"ios"=>NULL,"windows"=>"1"
(1 row)

SELECT multiply('ios=>1, windows=>1'::os_name_istore, 'android=>-1, windows=>1'::os_name_istore);
                  multiply                  
--------------------------------------------
 "android"=>NULL,"ios"=>NULL,"windows"=>"1"
(1 row)

SELECT multiply('android=>1, windows=>1'::os_name_istore, 'android=>-1, windows=>1'::os_name_istore);
            multiply            
--------------------------------
 "android"=>"-1","windows"=>"1"
(1 row)

SELECT multiply('android=>-1, windows=>1'::os_name_istore, 'android=>-1, windows=>1'::os_name_istore);
           multiply            
-------------------------------
 "android"=>"1","windows"=>"1"
(1 row)

SELECT multiply('android=>-1, windows=>1'::os_name_istore, 1);
            multiply            
--------------------------------
 "android"=>"-1","windows"=>"1"
(1 row)

SELECT multiply('android=>-1, windows=>1'::os_name_istore, -1);
            multiply            
--------------------------------
 "android"=>"1","windows"=>"-1"
(1 row)

SELECT multiply('android=>-1, windows=>1'::os_name_istore, 0);
           multiply            
-------------------------------
 "android"=>"0","windows"=>"0"
(1 row)

ROLLBACK;
BEGIN;
-- functions_os_name should generate an os_name_istore from array;
-- ./spec/functions_os_name_spec.rb:100;
CREATE EXTENSION istore;
SELECT os_name_istore_from_array(ARRAY['android']);
 os_name_istore_from_array 
---------------------------
 "android"=>"1"
(1 row)

SELECT os_name_istore_from_array(ARRAY['android','android','android','android']);
 os_name_istore_from_array 
---------------------------
 "android"=>"4"
(1 row)

SELECT os_name_istore_from_array(NULL::text[]);
 os_name_istore_from_array 
---------------------------
 
(1 row)

SELECT os_name_istore_from_array(ARRAY['android','ios','windows','windows-phone']);
                   os_name_istore_from_array                   
---------------------------------------------------------------
 "android"=>"1","ios"=>"1","windows"=>"1","windows-phone"=>"1"
(1 row)

SELECT os_name_istore_from_array(ARRAY['android','ios','windows','windows-phone','android','ios','windows','windows-phone']);
                   os_name_istore_from_array                   
---------------------------------------------------------------
 "android"=>"2","ios"=>"2","windows"=>"2","windows-phone"=>"2"
(1 row)

SELECT os_name_istore_from_array(ARRAY['android','ios','windows','windows-phone','android','ios','windows',NULL]);
                   os_name_istore_from_array                   
---------------------------------------------------------------
 "android"=>"2","ios"=>"2","windows"=>"2","windows-phone"=>"1"
(1 row)

SELECT os_name_istore_from_array(ARRAY[NULL,'ios','windows','windows-phone','android','ios','windows','windows-phone']);
                   os_name_istore_from_array                   
---------------------------------------------------------------
 "android"=>"1","ios"=>"2","windows"=>"2","windows-phone"=>"2"
(1 row)

SELECT os_name_istore_from_array(ARRAY[NULL,'ios','windows','windows-phone','android','ios','windows',NULL]);
                   os_name_istore_from_array                   
---------------------------------------------------------------
 "android"=>"1","ios"=>"2","windows"=>"2","windows-phone"=>"1"
(1 row)

SELECT os_name_istore_from_array(ARRAY['android','ios','windows',NULL,'android',NULL,'windows','windows-phone','android','ios','windows']);
                   os_name_istore_from_array                   
---------------------------------------------------------------
 "android"=>"3","ios"=>"2","windows"=>"3","windows-phone"=>"1"
(1 row)

SELECT os_name_istore_from_array(ARRAY['android'::os_name]);
 os_name_istore_from_array 
---------------------------
 "android"=>"1"
(1 row)

SELECT os_name_istore_from_array(ARRAY['android'::os_name,'android'::os_name,'android'::os_name,'android'::os_name]);
 os_name_istore_from_array 
---------------------------
 "android"=>"4"
(1 row)

SELECT os_name_istore_from_array(NULL::text[]);
 os_name_istore_from_array 
---------------------------
 
(1 row)

SELECT os_name_istore_from_array(ARRAY['android'::os_name,'ios'::os_name,'windows'::os_name,'windows-phone'::os_name]);
                   os_name_istore_from_array                   
---------------------------------------------------------------
 "android"=>"1","ios"=>"1","windows"=>"1","windows-phone"=>"1"
(1 row)

SELECT os_name_istore_from_array(ARRAY['android'::os_name,'ios'::os_name,'windows'::os_name,'windows-phone'::os_name,'android'::os_name,'ios'::os_name,'windows'::os_name,'windows-phone'::os_name]);
                   os_name_istore_from_array                   
---------------------------------------------------------------
 "android"=>"2","ios"=>"2","windows"=>"2","windows-phone"=>"2"
(1 row)

SELECT os_name_istore_from_array(ARRAY['android'::os_name,'ios'::os_name,'windows'::os_name,'windows-phone'::os_name,'android'::os_name,'ios'::os_name,'windows'::os_name,NULL]);
                   os_name_istore_from_array                   
---------------------------------------------------------------
 "android"=>"2","ios"=>"2","windows"=>"2","windows-phone"=>"1"
(1 row)

SELECT os_name_istore_from_array(ARRAY[NULL,'ios'::os_name,'windows'::os_name,'windows-phone'::os_name,'android'::os_name,'ios'::os_name,'windows'::os_name,'windows-phone'::os_name]);
                   os_name_istore_from_array                   
---------------------------------------------------------------
 "android"=>"1","ios"=>"2","windows"=>"2","windows-phone"=>"2"
(1 row)

SELECT os_name_istore_from_array(ARRAY[NULL,'ios'::os_name,'windows'::os_name,'windows-phone'::os_name,'android'::os_name,'ios'::os_name,'windows'::os_name, NULL]);
                   os_name_istore_from_array                   
---------------------------------------------------------------
 "android"=>"1","ios"=>"2","windows"=>"2","windows-phone"=>"1"
(1 row)

SELECT os_name_istore_from_array(ARRAY['android'::os_name,'ios'::os_name,'windows'::os_name,NULL,'android'::os_name,NULL,'windows'::os_name,'windows-phone'::os_name,'android'::os_name,'ios'::os_name,'windows'::os_name]);
                   os_name_istore_from_array                   
---------------------------------------------------------------
 "android"=>"3","ios"=>"2","windows"=>"3","windows-phone"=>"1"
(1 row)

ROLLBACK;
BEGIN;
-- functions_os_name should agg an array of os_name_istores;
-- ./spec/functions_os_name_spec.rb:154;
CREATE EXTENSION istore;
SELECT os_name_istore_agg(ARRAY['ios=>1']::os_name_istore[]);
 os_name_istore_agg 
--------------------
 "ios"=>"1"
(1 row)

SELECT os_name_istore_agg(ARRAY['ios=>1','ios=>1']::os_name_istore[]);
 os_name_istore_agg 
--------------------
 "ios"=>"2"
(1 row)

SELECT os_name_istore_agg(ARRAY['ios=>1,windows=>1','ios=>1,windows=>-1']::os_name_istore[]);
    os_name_istore_agg     
---------------------------
 "ios"=>"2","windows"=>"0"
(1 row)

SELECT os_name_istore_agg(ARRAY['ios=>1,windows=>1','ios=>1,windows=>-1',NULL]::os_name_istore[]);
    os_name_istore_agg     
---------------------------
 "ios"=>"2","windows"=>"0"
(1 row)

SELECT os_name_istore_agg(ARRAY[NULL,'ios=>1,windows=>1','ios=>1,windows=>-1']::os_name_istore[]);
    os_name_istore_agg     
---------------------------
 "ios"=>"2","windows"=>"0"
(1 row)

SELECT os_name_istore_agg(ARRAY[NULL,'ios=>1,windows=>1','ios=>1,windows=>-1',NULL]::os_name_istore[]);
    os_name_istore_agg     
---------------------------
 "ios"=>"2","windows"=>"0"
(1 row)

ROLLBACK;
BEGIN;
-- functions_os_name should sum_up os_name_istores;
-- ./spec/functions_os_name_spec.rb:174;
CREATE EXTENSION istore;
SELECT os_name_istore_sum_up('ios=>1'::os_name_istore);
 os_name_istore_sum_up 
-----------------------
                     1
(1 row)

SELECT os_name_istore_sum_up(NULL::os_name_istore);
 os_name_istore_sum_up 
-----------------------
                      
(1 row)

SELECT os_name_istore_sum_up('ios=>1, windows=>1'::os_name_istore);
 os_name_istore_sum_up 
-----------------------
                     2
(1 row)

SELECT os_name_istore_sum_up('ios=>1 ,windows=>-1, ios=>1'::os_name_istore);
 os_name_istore_sum_up 
-----------------------
                     1
(1 row)

ROLLBACK;
BEGIN;
-- functions_os_name should SUM os_names FROM table;
-- ./spec/functions_os_name_spec.rb:184;
CREATE EXTENSION istore;
CREATE TABLE test (a os_name_istore);
INSERT INTO test VALUES('ios=>1'),('windows=>1'),('windows-phone=>1');
SELECT SUM(a) FROM test;
                      sum                       
------------------------------------------------
 "ios"=>"1","windows"=>"1","windows-phone"=>"1"
(1 row)

ROLLBACK;
BEGIN;
-- functions_os_name should SUM os_names FROM table;
-- ./spec/functions_os_name_spec.rb:191;
CREATE EXTENSION istore;
CREATE TABLE test (a os_name_istore);
INSERT INTO test VALUES('ios=>1'),('windows=>1'),('windows-phone=>1'),(NULL),('windows-phone=>3');
SELECT SUM(a) FROM test;
                      sum                       
------------------------------------------------
 "ios"=>"1","windows"=>"1","windows-phone"=>"4"
(1 row)

ROLLBACK;
